### PR TITLE
Set the trove service password to random

### PR DESCRIPTION
**Pull request by s-t-e-v-e-n-k**
Force the trove service password to be random_password in the
model, to match other services.

### CODE REVIEW

**Comment by rhafer** [comment]
I think we should also create a schema migration in https://github.com/crowbar/crowbar-openstack/tree/master/chef/data_bags/crowbar/migrate/trove that, if the current pw is set to ```trove``` resets it to a random password.

**Comment by tboerger** [comment]
And the bot should exclude former team mates.

**Comment by s-t-e-v-e-n-k** [comment]
@rhafer I will investigate how feasible that is.

File: chef/data_bags/crowbar/template-trove.json [hunk 1]
```diff
@@ -25,7 +26,7 @@
   },
   "deployment": {
     "trove": {
-      "crowbar-revision": 1,
+      "crowbar-revision": 2,
       "crowbar-applied": false,
       "schema-revision": 100,
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/template-trove.json]: schema-revision needs to be bumped to 101

File: chef/data_bags/crowbar/template-trove.json [hunk 2]
```diff
@@ -25,7 +26,7 @@
   },
   "deployment": {
     "trove": {
-      "crowbar-revision": 1,
+      "crowbar-revision": 2,
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/template-trove.json]: that doesn't exist, it should stay with 1

File: chef/data_bags/crowbar/migrate/trove/101_add_service_user.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+def upgrade(ta, td, a, d)
+  a["service_user"] = ta["service_user"]
+  return a, d
+end
+
+def downgrade(ta, td, a, d)
+  a.delete("service_user")
+  return a, d
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/migrate/trove/101_add_service_user.rb]: service_password needs to be deleted as well

File: chef/data_bags/crowbar/migrate/trove/101_add_service_user.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+def upgrade(ta, td, a, d)
+  a["service_user"] = ta["service_user"]
+  return a, d
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/migrate/trove/101_add_service_user.rb]: service_password needs to be initialized to something sensible here

File: chef/data_bags/crowbar/migrate/trove/101_add_service_user.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+end
+
+def downgrade(ta, td, a, d)
+  a.delete("service_user")
```
Comment (dirkmueller) [inline chef/data_bags/crowbar/migrate/trove/101_add_service_user.rb]: do we need to back port this to cloud 6 as this smells like a security bug?

**Comment by rhafer** [comment]
Please add the bsc reference to the summary line of the commit message. That way our scripts will catch it for the changelog of the RPM packages.

**Comment by rhafer** [comment]
BTW, for the generation of the password in the schema migration feel free to get some inspiration from: https://github.com/crowbar/crowbar-openstack/blob/master/chef/data_bags/crowbar/migrate/glance/004_add_db.rb

It's a little tricky since the migration is usually run two times on systems where trove is already deployed (1st for the Proposal, stored in crowbar's datbase, 2nd for the chef-role that holds the currently deployed configuration). The password needs to be set to the same value for both runs, that's why in class variable is needed there. (I think you can ignore the stuff dealing with node attribute from the above reference migration)

File: chef/data_bags/crowbar/migrate/trove/101_add_service_user.rb (hunk 5 above)
Comment (rhafer) [inline chef/data_bags/crowbar/migrate/trove/101_add_service_user.rb]: > do we need to back port this to cloud 6 as this smells like a security bug?

Yes, definitely.

**Comment by dirkmueller** [comment]
@s-t-e-v-e-n-k ping? :)

File: chef/data_bags/crowbar/template-trove.schema [hunk 6]
```diff
@@ -20,6 +20,8 @@
             "swift_instance": { "type": "str", "required": true },
             "rabbitmq_instance": { "type": "str", "required": true },
             "volume_support": { "type": "bool", "required": true },
+            "service_user": { "type": "str", "required": true },
+            "service_password": { "type": "str" },
```
Comment (Itxaka) [inline chef/data_bags/crowbar/template-trove.schema]: service_password in on the schema but not on the json. Shouldn't it be?

**Comment by Itxaka** [comment]
Apart from the small nit, looks good.

**Comment by Itxaka** [comment]
retriggering jenkins as it seems that the job never posted back :/

**Comment by nicolasbock** [comment]
+1

File: chef/data_bags/crowbar/migrate/trove/101_add_service_details.rb [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  unless defined?(@@trove_service_password)
+    service = ServiceObject.new "fake-logger"
+    @@trove_service_password = service.random_password
+  end
+  Chef::Search::Query.new.search(:node) do |node|
+    unless defined?(node[:trove])
+      node[:trove][:service_password] = @@trove_service_password
+      node.save
+    end
+  end
```
Comment (vuntz) [inline chef/data_bags/crowbar/migrate/trove/101_add_service_details.rb]: I don't think we need to change the nodes; the migration will change the role for the proposals, so the nodes will automatically get the attribute.

File: chef/data_bags/crowbar/template-trove.schema (hunk 6 above)
Comment (s-t-e-v-e-n-k) [inline chef/data_bags/crowbar/template-trove.schema]: We can't set it to a reasonable value in the JSON, since we want it to a random password.

**Comment by s-t-e-v-e-n-k** [comment]
Thanks for the review @vuntz. You're right that the node setting isn't required. I added the code since during my testing, I saw the updated password in the crowbar database, but not in the output of crowbar machine show. I've re-tested it without that chunk of code, and confirmed that it does show up in crowbar machine show after chef-client has run, so I've removed it.

**Comment by vuntz** [comment]
Thanks, +1

**Comment by Itxaka** [comment]
Looks good, thanks for addressing my comment +1

**Comment by Itxaka** [comment]
retriggered the job :)

**Pull request merged**